src/types/redis_hash.cc

```diff
@@ -386,5 +387,78 @@ rocksdb::Status Hash::Scan(const Slice &user_key, const std::string &cursor, uin
   return SubKeyScanner::Scan(kRedisHash, user_key, cursor, limit, field_prefix, fields, values);
 }
+rocksdb::Status Hash::RandField(const Slice &user_key, std::vector<FieldValue> *field_values, uint64_t count, bool uniq,
+                                HashFetchType type) {
+  field_values->clear();
+  std::string ns_key;
+  AppendNamespacePrefix(user_key, &ns_key);
+  HashMetadata metadata(false);
+  rocksdb::Status s = GetMetadata(ns_key, &metadata);
+  if (!s.ok()) return s;
+  std::string prefix_key, next_version_prefix_key;
+  InternalKey(ns_key, "", metadata.version, storage_->IsSlotIdEncoded()).Encode(&prefix_key);
+  InternalKey(ns_key, "", metadata.version + 1, storage_->IsSlotIdEncoded()).Encode(&next_version_prefix_key);
 
+  rocksdb::ReadOptions read_options;
+  LatestSnapShot ss(storage_);
+  read_options.snapshot = ss.GetSnapShot();
+  rocksdb::Slice upper_bound(next_version_prefix_key);
+  read_options.iterate_upper_bound = &upper_bound;
+  storage_->SetReadOptions(read_options);
+
+  uint64_t size = metadata.size;
+  std::vector<std::string> fields;
+  std::vector<std::string> values;
+  auto iter = util::UniqueIterator(storage_, read_options);
+  for (iter->Seek(prefix_key); iter->Valid() && iter->key().starts_with(prefix_key); iter->Next()) {
+    InternalKey ikey(iter->key(), storage_->IsSlotIdEncoded());
+    fields.emplace_back(ikey.GetSubKey().ToString());
+    if (type == HashFetchType::kAll) {
+      values.emplace_back(iter->value().ToString());
+    }
+  }
+
+  // Case 1: Negative count, randomly select elements
+  if (!uniq) {
+    field_values->reserve(count);
+    for (uint64_t i = 0; i < count; i++) {
+      uint64_t index = (std::rand() % size);
+      if (type == HashFetchType::kAll) {
+        field_values->emplace_back(fields[index], values[index]);
+      } else {
+        field_values->emplace_back(fields[index], "");
+      }
+    }
+  }
+
+  // Case 2: Requested count is greater than or equal to the number of elements inside the hash
+  else if (size <= count) {
+    field_values->reserve(size);
+    for (uint64_t i = 0; i < size; i++) {
+      if (type == HashFetchType::kAll) {
+        field_values->emplace_back(fields[i], values[i]);
+      } else {
+        field_values->emplace_back(fields[i], "");
+      }
+    }
+  }
+  // Case 3: Requested count is less than the number of elements inside the hash
+  else {
+    field_values->reserve(count);
+    std::vector<uint64_t> indices(size);
+    std::iota(indices.begin(), indices.end(), 0);
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    std::shuffle(indices.begin(), indices.end(), gen);  // use Fisher-Yates shuffle algorithm to randomize the order
+    for (uint64_t i = 0; i < count; i++) {
+      uint64_t index = indices[i];
+      if (type == HashFetchType::kAll) {
+        field_values->emplace_back(fields[index], values[index]);
+      } else {
+        field_values->emplace_back(fields[index], "");
+      }
+    }
+  }
```

Comment:
seems the output will always be a random order of index {0,...count-1}, which cannot be called "rand field" IMHO.

Reply:
if you want to do the shuffle, you need to shuffle [0, `hash_elem_size`) and cut the first `count` number as indices.

Reply:
> if you want to do the shuffle, you need to shuffle [0, `hash_elem_size`) and cut the first `count` numbers as indices.

should I modify it like this, or what should I do
~~~c
std::shuffle(indices.begin(), indices.begin() + count, gen);  
std::sort(indices.begin(), indices.begin() + count);
~~~

Reply:
> if you want to do the shuffle, you need to shuffle [0, `hash_elem_size`) and cut the first `count` numbers as indices.
should I modify like this,or anything else
~~~cpp
indices.resize(count);
~~~

Reply:
> if you want to do the shuffle, you need to shuffle [0, `hash_elem_size`) and cut the first `count` numbers as indices.

should I modify like this,or anything else
~~~cpp
indices.resize(count);
~~~

Reply:
oh I may mistake `std::vector indices(size)` as `indices(count)`, sorry for that. the implementation looks fine to me.

Reply:
I think you can just use random_device as the URBG without mt19937.

Reply:
A little reference for [why not just use random_device](https://stackoverflow.com/questions/39288595/why-not-just-use-random-device).  :)

Reply:
usally random device can be used to generate a mt19937 seed

Reply:
> A little reference for [why not just use random_device](https://stackoverflow.com/questions/39288595/why-not-just-use-random-device). :)

@maplefu @Yangsx-1  I am not sure the common way to implement std::random_device is still `/dev/urandom`, the answer seems to be a little out dated.

Usually, e.g. in x86, the `rdrand` instruction is a far better way to implement std::random_device with great performance.

Reply:
> usally random device can be used to generate a mt19937 seed

mt19937 is just a good PRNG algorithm. it is not required to use if your device can generate **real random numbers**.

Reply:
@Yangsx-1 I have checked the latest version of libstdc++, which are using `rdrand` when possible.

https://github.com/gcc-mirror/gcc/blob/63ae6bc60c0f67fb2791991bf4b6e7e0a907d420/libstdc%2B%2B-v3/src/c%2B%2B11/random.cc#L34

Reply:
```
/* If our unsigned long type can store a 64 bit number, use a 64 bit PRNG. */
#if ULONG_MAX >= 0xffffffffffffffff
#define randomULong() ((unsigned long) genrand64_int64())
#else
#define randomULong() random()
#endif
```

@PragmaTwice after check in redis, I guess it would uses `randomULong()`, and it wraps a PRNG algorithm. I guess mt19937 would be ok